### DISCUSSION TITLE
How to connect multiple m-n relationship records

### DISCUSSION THREAD

**Discussion by tomaszferens** [prisma/prisma#6766] [Q&A] [answered]
I have a schema that looks like this:

```prisma
model School {
  id       Int       @id @default(autoincrement())
  name     String
  teachers Teacher[]
  students Student[]
}

model Teacher {
  id        Int       @id @default(autoincrement())
  firstName String
  lastName  String
  schools   School[]
  students  Student[]
}

model Student {
  id        Int     @id @default(autoincrement())
  firstName String
  lastName  String
  teacherId Int
  teacher   Teacher @relation(fields: [teacherId], references: [id])
  schoolId  Int
  school    School  @relation(fields: [schoolId], references: [id])
}
```

I want to create in a single call a school, with teacher and student. This is my attempt:

```ts
prisma.school.create({
    data: {
        name: "Best",
        teachers: {
            create: {
                firstName: "Charles",
                lastName: "Wise",
                students: {
                    create: {
                        firstName: "Pauline",
                        lastName: "Jenkins",
                        school: {
                            connect: {
                                id: -1 // How to reference a school that is being created?
                            }
                        }
                    }
                }
            }
        }
    }
});
```

Unfortunately I have no clue how to pass school's id when connection it with a student.

I know that I can do it in two calls by first creating a school with a teacher and then creating a student with that just created school and teacher:

```ts
const school = await prisma.school.create({
    data: {
        name: "Best",
        teachers: {
            create: {
                firstName: "Charles",
                lastName: "Wise",
            }
        }
    },
    include: {
        teachers: true,
    }
});
await prisma.student.create({
    data: {
        firstName: "Pauline",
        lastName: "Jenkins",
        school: {
            connect: {
                id: school.id
            }
        },
        teacher: {
            connect: {
                id: school.teachers[0].id
            }
        }
    }
});
```
But that looks kind of weird and error prone. 

Is there a way to rewrite it to a single call? 🤔

**Comment by ruheni** [comment]
Hi @tomaszferens 👋🏾

Alex here from the Prisma Team.

This won't be possible given the "cyclical" nature of the relationships. You can add an if statement around the create student query to capture the school id before creating a student:
```ts
const school = await prisma.school.create({
  data: {
    name: "Best",
    teachers: {
      create: {
        firstName: "Charles",
        lastName: "Wise",
      }
    }
  },
  include: { teachers: true }
})

if (school.id) {
  const student = await prisma.student.create({
    data: {
      firstName: 'Pauline',
      lastName: 'Jenkins',
      schoolId: school.id,
      teacherId: school.teachers[0].id
    }
  })
}
```
To handle the errors, you can wrap your queries around a  `try{ } catch ( ){ }` block to capture any errors that come up.

**Reply by tomaszferens** [reply, marked as answer]
Thanks for the answer @ruheni.

I actually "fixed" it by pregenerating school's ID before creating a school. I had to change the id to be a String instead of an Int.

```ts
import cuid from 'cuid';

const schoolId = cuid();
prisma.school.create({
    data: {
        id: schoolId,
        name: "Best",
        teachers: {
            create: {
                firstName: "Charles",
                lastName: "Wise",
                students: {
                    create: {
                        firstName: "Pauline",
                        lastName: "Jenkins",
                        school: {
                            connect: {
                                id: schoolId,
                            }
                        }
                    }
                }
            }
        }
    }
});
```

This way it's a single call and I don't have to deal with rollbacks when one of the two calls fails (like in your example; creating a student might fail - does it mean that I should delete a school and a teacher created previously?).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
